tasks/resolve_reference.rb

```diff
@@ -0,0 +1,200 @@
+
+class GCloudInventory < TaskHelper
+  include RubyPluginHelper
+
+  AUTH_SCOPE        = 'https://www.googleapis.com/auth/compute.readonly'
+  AUTH_SKEW         = 60
+  CREDENTIALS_ENV   = 'GOOGLE_APPLICATION_CREDENTIALS'
+  CREDENTIALS_KEYS  = %w[client_email private_key project_id token_uri].freeze
+  GRANT_TYPE        = 'urn:ietf:params:oauth:grant-type:jwt-bearer'
+  SIGNING_ALGORITHM = 'RS256'
+
+  def resolve_reference(opts)
+    # Validate the target mapping
+    template = opts.delete(:target_mapping) || {}
+    unless template.key?(:uri) || template.key?(:name)
+      msg = "You must provide a 'name' or 'uri' in 'target_mapping' for the Google Cloud inventory plugin"
+      raise TaskHelper::Error.new(msg, 'bolt.plugin/validation-error')
+    end
+
+    # Retrieve credentials and access token for the API requests
+    creds = credentials(opts)
+    token = token(creds)
+    url   = "https://compute.googleapis.com/compute/v1/projects/#{opts[:project]}/zones/#{opts[:zone]}/instances"
+
+    # Build a list of compute engine instances, making multiple API requests if needed
+    instances = get_all_instances(url, token)
+
+    # Pull the necessary target data from each instance
+    attributes = required_data(template)
+    target_data = instances.map do |instance|
+      attributes.each_with_object({}) do |attr, acc|
+        attr = attr.first
+        acc[attr] = instance.key?(attr) ? instance[attr] : nil
+      end
+    end
+
+    # Apply the target data to the template
+    target_data.map { |data| apply_mapping(template, data) }
+  end
+
+  # Hash of required credentials for authorizing with Google authentication server
+  # The file path to the credentials file is loaded from either the plugin config
+  # or from an evironment variable
+  def credentials(opts)
+    # Ensure a credentials file was specified
+    unless opts[:credentials] || ENV[CREDENTIALS_ENV]
+      msg = "Missing application credentials. Specify the path to the application credentials file "\
+            "under the 'credentials' configuration option or as the 'GOOGLE_APPLICATION_CREDENTIALS' "\
+            "environment variable."
+      raise TaskHelper::Error.new(msg, 'bolt.plugin/validation-error')
+    end
+
+    path        = File.expand_path(opts[:credentials] || ENV[CREDENTIALS_ENV])
+    credentials = JSON.parse(File.read(path))
+
+    # Ensure the credentials are a hash
+    unless credentials.is_a? Hash
+      msg = "Expected credentials to be a Hash, received #{credentials.class}"
+      raise TaskHelper::Error.new(msg, 'bolt.plugin/validation-error')
+    end
+
+    # Ensure the credentials have the required keys
+    if (keys = CREDENTIALS_KEYS - credentials.keys).any?
+      msg = "Missing required keys in credentials file: #{keys.join(', ')}"
+      raise TaskHelper::Error.new(msg, 'bolt.plugin/validation-error')
+    end
+
+    # Ensure the credentials are for the specified project
+    unless credentials['project_id'] == opts[:project]
```

Comment:
Why do you need to specify `project` if it's in the credentials file?

Reply:
Removed the part that checks `project_id` in the credentials file. Not all credentials files will have this field, so the `project` field should still be required.

Fyi, something wonky happened with the git history when I pushed changes and it auto-closed the PR. I've got a new one up here: #4